core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTest.java

```diff
@@ -0,0 +1,136 @@
+            query.must(createRandomQuery());
+        }
+        for (int i = 0; i < randomIntBetween(0, 3); i++) {
+            query.mustNot(createRandomQuery());
+        }
+        for (int i = 0; i < randomIntBetween(0, 3); i++) {
+            query.should(createRandomQuery());
+        }
+        for (int i = 0; i < randomIntBetween(0, 3); i++) {
+            query.filter(createRandomQuery());
+        }
+        if (randomBoolean()) {
+            query.queryName(randomUnicodeOfLengthBetween(3, 15));
+        }
+        return query;
+    }
+
+    private static QueryBuilder createRandomQuery() {
+        QueryBuilder query = null;
+        switch (randomIntBetween(0, 2)) {
+        case 0:
+            return new MatchAllQueryBuilderTest().createTestQueryBuilder();
+        case 1:
+            return new TermQueryBuilderTest().createTestQueryBuilder();
+        case 2:
+            return new IdsQueryBuilderTest().createTestQueryBuilder();
+        }
+        return query;
+    }
+
+    @Override
+    protected BoolQueryBuilder createEmptyQueryBuilder() {
+        return new BoolQueryBuilder();
+    }
+
+    @Override
+    protected Query createExpectedQuery(BoolQueryBuilder queryBuilder, QueryParseContext context) throws IOException {
+        if (!queryBuilder.hasClauses()) {
+            return new MatchAllDocsQuery();
+        }
+
+        BooleanQuery boolQuery = new BooleanQuery(queryBuilder.disableCoord());
+        boolQuery.setBoost(queryBuilder.boost());
+        addBooleanClauses(context, boolQuery, queryBuilder.must(), BooleanClause.Occur.MUST);
+        addBooleanClauses(context, boolQuery, queryBuilder.mustNot(), BooleanClause.Occur.MUST_NOT);
+        addBooleanClauses(context, boolQuery, queryBuilder.should(), BooleanClause.Occur.SHOULD);
+        addBooleanClauses(context, boolQuery, queryBuilder.filter(), BooleanClause.Occur.FILTER);
+
+        Queries.applyMinimumShouldMatch(boolQuery, queryBuilder.minimumNumberShouldMatch());
+        Query returnedQuery = queryBuilder.adjustPureNegative() ? fixNegativeQueryIfNeeded(boolQuery) : boolQuery;
+        return returnedQuery;
+    }
+
+    private void addBooleanClauses(QueryParseContext parseContext, BooleanQuery booleanQuery, List<QueryBuilder> clauses, Occur occurs)
+            throws IOException {
+        for (QueryBuilder query : clauses) {
+            Query luceneQuery = null;
+            if (query instanceof MatchAllQueryBuilder) {
+                luceneQuery = new MatchAllQueryBuilderTest().createExpectedQuery((MatchAllQueryBuilder) query, parseContext);
+            }
+            if (query instanceof TermQueryBuilder) {
+                luceneQuery = new TermQueryBuilderTest().createExpectedQuery((TermQueryBuilder) query, parseContext);
+            }
+            if (query instanceof IdsQueryBuilder) {
+                luceneQuery = new IdsQueryBuilderTest().createExpectedQuery((IdsQueryBuilder) query, parseContext);
```

Comment:
Rather than setting up a whole new system here I decided to rely on toQuery() for all nested queries (after all we test the main QueryBuilder here) in creating the expected BoolQuery and not use createExpectedQuery here. Let me know what you think, we can alway extend the scope here.